### PR TITLE
Add a ALTER CLUSTER SWAP TABLE statement

### PR DESCRIPTION
**Pull request by mfussenegger**
- [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: blackbox/docs/appendices/release-notes/unreleased.rst [hunk 1]
```diff
@@ -46,6 +46,9 @@ Breaking Changes
 Changes
 =======
 
+- Added a ``ALTER CLUSTER SWAP TABLE`` statement that can be used to switch two
+  tables.
```
Comment (seut) [inline blackbox/docs/appendices/release-notes/unreleased.rst]: maybe more explicit: "... switch two table names."

**Comment by mfussenegger** [comment]
Also fyi: @joemoe

File: sql/src/main/java/io/crate/execution/ddl/TransportSwapRelationsAction.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                protected AcknowledgedResponse newResponse(boolean acknowledged) {
+                    return new AcknowledgedResponse(acknowledged);
+                }
+            };
+        clusterService.submitStateUpdateTask("swap-relations", updateTask);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(SwapRelationsRequest request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/ddl/TransportSwapRelationsAction.java]: we should also take index level blocks into account

File: enterprise/users/src/main/java/io/crate/auth/user/UserManagerDDLModifier.java [hunk 3]
```diff
@@ -76,22 +91,24 @@ private static boolean dropPrivileges(MetaData.Builder mdBuilder, RelationName r
         return affectedRows > 0L;
     }
 
-    private static boolean transferTablePrivileges(MetaData.Builder mdBuilder,
-                                                RelationName sourceRelationName,
-                                                RelationName targetRelationName) {
-        UsersPrivilegesMetaData oldMetaData = (UsersPrivilegesMetaData) mdBuilder.getCustom(UsersPrivilegesMetaData.TYPE);
-        if (oldMetaData == null) {
+    private static boolean transferTablePrivileges(MetaData previousMetaData,
```
[the remark this answers is not in the document]
Comment (seut) [reply enterprise/users/src/main/java/io/crate/auth/user/UserManagerDDLModifier.java]: Was wondering first as well. I don't mind, I like your changes (maybe a dedicated commit?).

File: sql/src/main/java/io/crate/planner/DependencyCarrier.java [hunk 4]
```diff
@@ -58,6 +59,7 @@
     private final ProjectionBuilder projectionBuilder;
     private final TransportCreateViewAction createViewAction;
     private final TransportDropViewAction dropViewAction;
+    private final TransportSwapRelationsAction alterIndicesAction;
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/DependencyCarrier.java]: I'd prefer `swapRelationsAction`.

File: blackbox/docs/sql/statements/alter-cluster.rst [hunk 5]
```diff
@@ -19,6 +19,7 @@ Synopsis
 
     ALTER CLUSTER
       { REROUTE RETRY FAILED }
+      | { SWAP TABLE source TO target [ WITH ( expr = expr [ , ... ] ) ] }
```
[the remark this answers is not in the document]
Comment (seut) [reply blackbox/docs/sql/statements/alter-cluster.rst]: hm I like it.

[the remark this answers is not in the document]
Comment (smakalias) [reply blackbox/docs/sql/statements/alter-cluster.rst]: I think we either say "swap something *with* somethingElse" or ""swap something *for* somethingElse".

If you're keeping SWAP, maybe consider this

[the remark this answers is not in the document]
Comment (mfussenegger) [reply blackbox/docs/sql/statements/alter-cluster.rst]: I didn't want to duplicate `WITH` as it's already used for the options. `FOR` sounds also strange to me as non-native, that's why I thought of omitting it completely.

File: sql/src/test/java/io/crate/analyze/SwapTableAnalyzerTest.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    @Test
+    public void testSwapTableStatementCanBeAnalysedIfBothTablesExist() throws Exception {
+        AnalyzedSwapTable statement = e.analyze("alter cluster swap table t1 to t2 with (drop_source = true)");
+        assertThat(statement.source().toString(), is("doc.t1"));
+        assertThat(statement.target().toString(), is("doc.t2"));
+        assertThat(statement.dropSource(), isLiteral(true));
+    }
+
+    @Test
+    public void testPruneSourceDefaultsToFalse() {
```
Comment (smakalias) [inline sql/src/test/java/io/crate/analyze/SwapTableAnalyzerTest.java]: testDropSource...

**Comment by smakalias** [comment]
One more general comment. We could potentially de-corralate  concreteIndex `Index` handling from Crate specific notions such as tables partitions etc. 
Reasons for doing so:
a) we call `concreteIndeces` only once for `source` and `target` 
b) it can be reused as is in the shrinking Operation

Just a suggestion, it can be done as separate task later on, I just realise that most code can and should be reused for the shrink operation.

File: enterprise/users/src/main/java/io/crate/auth/user/UserManagerDDLModifier.java [hunk 7]
```diff
@@ -54,6 +54,21 @@ public ClusterState onRenameTable(ClusterState currentState,
...
     }
 
+    @Override
+    public ClusterState onSwapRelations(ClusterState currentState, RelationName source, RelationName target) {
+        MetaData currentMetaData = currentState.metaData();
+        UsersPrivilegesMetaData userPrivileges = currentMetaData.custom(UsersPrivilegesMetaData.TYPE);
+        if (userPrivileges == null) {
+            return currentState;
+        }
+        UsersPrivilegesMetaData updatedPrivileges = UsersPrivilegesMetaData.swapPrivileges(userPrivileges, source, target);
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply enterprise/users/src/main/java/io/crate/auth/user/UserManagerDDLModifier.java]: With the data structure of the privileges it is not possible to simply drop and copy.

File: sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java [hunk 8]
```diff
@@ -123,6 +124,12 @@ public void testDropTableStmtBuilder() {
         printStatement("drop table bar.foo");
     }
 
+    @Test
+    public void testSwapTable() {
+        printStatement("ALTER CLUSTER SWAP TABLE t1 TO t2");
+        printStatement("ALTER CLUSTER SWAP TABLE t1 TO t2 WITH (prune_second = true)");
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java]: yep, but doesn't really matter for the parser test as it is unaware of the allowed options.

File: sql/src/main/java/io/crate/execution/ddl/SwapRelationsOperation.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        List<RelationName> dropRelations = swapRelationsRequest.dropRelations();
+        if (dropRelations.isEmpty()) {
+            return stateAfterRename;
+        } else {
+            return applyDropRelations(state, stateAfterRename, dropRelations);
+        }
+    }
+
+    private UpdatedState applyDropRelations(ClusterState state, UpdatedState updatedState, List<RelationName> dropRelations) {
+        ClusterState stateAfterRename = updatedState.newState;
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/ddl/SwapRelationsOperation.java]: I agree that functions should generally not be aware of the context they're used in, but I think here it helps more with readability than it hurts, so I'd keep it as is

**Comment by mfussenegger** [comment]
@seut pushed a fixup, I figured that relying on there being at least 1 index to determine the partitioning breaks for empty partitioned tables.

**Pull request merged**